### ISSUE TITLE
Creating content item(s) from Workflow issue

### ISSUE THREAD

**Issue by Skrypt** [OrchardCMS/OrchardCore#4886]
When we draft or publish a content item from the admin UI we index(map) in database the AutoroutePart with the AutoroutePartIndexProvider. It gets saved in the AutoroutePartIndex table.
This works fine.

Though if we try to create a content item from a Workflow the AutoroutePartIndexProvider will fail to retrieve the proper value of the AutoroutePart.Path.

```
var path = contentItem.As<AutoroutePart>()?.Path;
```

This line takes the contentItem.Elements and parse it to create and return an actual AutoroutePart object. Though in the context of a Workflow execution, this ContentItem.Elements["AutoroutePart"] will return a JObject with a Path param = null.

This results in not getting the AutoroutePart to get Indexed as the Path always = null in the context of a Workflow execution. Same issue can be found with the AliasPart. Though the ContentFieldIndexes are getting executed fine.

**Comment by Skrypt** [comment]
Here some temp fix : 

```
using System;
using Newtonsoft.Json.Linq;
using OrchardCore.Autoroute.Models;
using YesSql.Indexes;

namespace OrchardCore.ContentManagement.Records
{
    public class AutoroutePartIndex : MapIndex
    {
        public string ContentItemId { get; set; }
        public string Path { get; set; }
        public bool Published { get; set; }
    }

    public class AutoroutePartIndexProvider : IndexProvider<ContentItem>
    {
        public override void Describe(DescribeContext<ContentItem> context)
        {
            context.For<AutoroutePartIndex>()
                .Map(contentItem =>
                {
                    //var path = contentItem.As<AutoroutePart>()?.Path;
                    var content = (JObject)contentItem.Content;
                    var autoroutePart = content["AutoroutePart"]?.ToObject<AutoroutePart>();
                    var path = autoroutePart.Path;

                    if (path == null) {
                        var json = contentItem.Content;
                        var part = json["AutoroutePart"];
                        path = part["Path"];
                    }

                    if (!String.IsNullOrEmpty(path) && (contentItem.Published || contentItem.Latest))
                    {
                        return new AutoroutePartIndex
                        {
                            ContentItemId = contentItem.ContentItemId,
                            Path = path,
                            Published = contentItem.Published
                        };
                    }

                    return null;
                });
        }
    }
}
```

**Comment by mawendel** [comment]
Will the hack be implementet temporarily or are we waiting for the real fix?

**Comment by Skrypt** [comment]
This will get triaged on thursday. I need opinions of other people on that one as this looks like something that should come from YesSql. I'm currently not sure what affects `contentItem.Elements`.

**Comment by deanmarcussen** [comment]
I looked quickly at this.

the problem is in https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Modules/OrchardCore.Contents/Workflows/Activities/CreateContentTask.cs

```
           if (!string.IsNullOrWhiteSpace(ContentProperties.Expression))
            {
                var contentProperties = await _expressionEvaluator.EvaluateAsync(ContentProperties, workflowContext);
                var propertyObject = JObject.Parse(contentProperties);

                ((JObject)contentItem.Content).Merge(propertyObject);
            }
```

that code needs to call `contentItem.Apply()`

probably the fix is something like
```
    var propertyObject = JObject.Parse(contentProperties);
    var contentElement = propertyObject.ToObject<ContentElement>();
    contentItem.Apply(contentElement);
```

but I am not sure that it will cast to ContentElement, or ContentItem. 

Will try and have a look but short of time this week

**Comment by Skrypt** [comment]
Could make sense.

**Comment by Skrypt** [comment]
Ok just tested this and it still doesn't work. Here is why.
The DisplayText property of a contentItem is accessible by doing contentItem.DisplayText.
((JObject)contentItem.Content).Merge(propertyObject); merges the JSON that we set in the workflow create task and adds a DisplayText inside contentItem.Content which is not what it should do. I believe this is because we used to have a TitlePart that could be modified easily using contentItem.Content.TitlePart.
